packages/zoltan2/src/algorithms/connectivity/Zoltan2_VtxLabel.hpp

```diff
@@ -0,0 +1,713 @@
+#include "Tpetra_Core.hpp"
+#include "Teuchos_RCP.hpp"
+#include "Tpetra_Import.hpp"
+#include "Tpetra_FEMultiVector.hpp"
+
+#include <string>
+#include <sstream>
+#include <iostream>
+#include <queue>
+
+#ifndef ZOLTAN2_VTXLABEL_
+#define ZOLTAN2_VTXLABEL_
+
+namespace Zoltan2{
+        
+	//Graph datastructure that represents the local
+	//graph in ice sheet propagation problems
+        template<typename lno_t>
+	class icePropGraph {
+          public:
+            size_t n;
+            size_t m;
+            lno_t* out_array;
+            lno_t* out_degree_list;
+            lno_t max_degree_vert;
+            double avg_out_degree;
+            lno_t out_degree(lno_t vert){
+              return (out_degree_list[vert+1] - out_degree_list[vert]);
+            }
+            lno_t* out_vertices(lno_t vert){
+              return (&out_array[out_degree_list[vert]]);
+            }
+        }; 
+        
+        //global queues to allow easy propagation from the
+	//+= operator of IcePropVtxLabel.
+	std::queue<int> icePropArtQueue;
+	std::queue<int> icePropRegQueue;
+
+        //Enum that denotes
+	enum IcePropGrounding_Status {ICEPROPGS_FULL=2, ICEPROPGS_HALF=1, ICEPROPGS_NONE = 0};
+	// Struct representing a vertex label.
+	// We define our own "addition" for these labels.
+	// Later, we'll create a Tpetra::FEMultiVector of these labels.
+	class IcePropVtxLabel {
```

Comment:
All fields of data structure need comments.  What are they?

Reply:
I was surprised that the graph is templated on lno_t, but the vertex labels are not.
Aren't the senders and labels actually vertex IDs?  In fact, I thought they were global vertex IDs.  If so, they should be defined with the templated type gno_t.

Reply:
This was a similar issue to the return type of the flags. I use a sentinel value to indicate an empty label, this class should be templated on gno_t. In the spirit of the other solution, a way around the sentinel value could be to add two fields indicating whether or not a label has a valid value in it.

Reply:
I use -1 to represent unused currently. The labels contain global identifiers, so they should be templated on the global ordinal type.

Reply:
I'm currently reworking these last changes, I have templated the label class on both gno_t and lno_t, because the id field is actually a local identifier.